src/validator.cpp

```diff
@@ -1376,4 +1425,237 @@ void Validator::ValidatorImpl::checkUnitForCycles(const ModelPtr &model, const U
+                if ((s1 < s2) && !connectionId.empty() && (reportedConnections.count(connection) == 0)) {
+                    reportedConnections.insert(connection);
+                    info = " - connection between components '" + component->name() + "' and '" + equivParent->name()
+                           + "' because of variable equivalence between variables '" + item->name()
+                           + "' and '" + equiv->name() + "'";
+                    addIdMapItem(connectionId, info, idMap);
+                }
+            }
+        }
+    }
+
+    // Resets.
+    for (size_t i = 0; i < component->resetCount(); ++i) {
+        auto item = component->reset(i);
+        if (!item->id().empty()) {
+            info = " - reset at index " + std::to_string(i) + " in component '" + component->name() + "'";
+            addIdMapItem(item->id(), info, idMap);
+        }
+        if (!item->testValueId().empty()) {
+            info = " - test_value in reset at index " + std::to_string(i) + " in component '" + component->name() + "'";
+            addIdMapItem(item->testValueId(), info, idMap);
+        }
+        info = "test_value in reset " + std::to_string(i) + " in component '" + component->name() + "'";
+        buildMathIdMap(info, idMap, item->testValue());
+        if (!item->resetValueId().empty()) {
+            info = " - reset_value in reset at index " + std::to_string(i) + " in component '" + component->name() + "'";
+            addIdMapItem(item->resetValueId(), info, idMap);
+        }
+        info = "reset_value in reset " + std::to_string(i) + " in component '" + component->name() + "'";
+        buildMathIdMap(info, idMap, item->resetValue());
+    }
+
+    // Maths.
+    info = "math in component '" + component->name() + "'";
+    buildMathIdMap(info, idMap, component->math());
+
+    // Imports.
+    if (component->isImport() && (component->importSource() != nullptr) && !component->importSource()->id().empty()) {
+        info = " - import source for component '" + component->name() + "'";
+        addIdMapItem(component->importSource()->id(), info, idMap);
+    }
+
+    // Connections.
+    if (!component->encapsulationId().empty()) {
+        info = " - encapsulation component_ref to component '" + component->name() + "'";
+        addIdMapItem(component->encapsulationId(), info, idMap);
+    }
+
+    // Child components.
+    for (size_t c = 0; c < component->componentCount(); ++c) {
+        buildComponentIdMap(component->component(c), idMap, reportedConnections);
+    }
+}
+
+void Validator::ValidatorImpl::buildMathIdMap(const std::string &infoRef, IdMap &idMap, const std::string &input)
+{
+    std::vector<XmlDocPtr> docs = multiRootXml(input);
+
+    for (const auto &doc : docs) {
+        XmlNodePtr node = doc->rootNode();
+        if (node == nullptr) {
+            return;
+        }
+        if (!node->isMathmlElement("math")) {
+            return;
```

Comment:
why is this return instead of ignoring non-math root nodes?

Reply:
7f063dfa